### PR TITLE
fix(config): handle query config options

### PR DESCRIPTION
**Pull request by glowcloud**
Refs #9807

### CODE REVIEW

File: src/core/config/merge.js [hunk 1]
```diff
@@ -34,6 +35,13 @@ const merge = (target, ...sources) => {
     merged.domNode = domNode
   }
 
+  Object.entries(merged).forEach(([key, value]) => {
```
Comment (glowcloud) [inline src/core/config/merge.js:38]: I put this change here instead of in `query` because of `urls.primaryName`. 

The issue is that if we set `urls.primaryName`, we’ll have 
```js
config: { urls: { primaryName: 'string' } }
```
 so when using deep extend where in config of lower priority we'll have:
```js
config: { urls: [{ url: 'url1', name: 'string' }] }
```
The result of this will just be the object:
```js
config: { urls: { primaryName: 'string' } }
```

I tried setting it by doing something like
```js
queryOptions.urls = []
set(queryOptions, key, value)
```

Which seemed fine, as we had an empty array with `primaryName` property, but then when merging it with the other options, I would just get an empty array. 
A quick test for this:
```js
  const foo = { urls: [] }
  set(foo, "urls.primaryName", "string")
  const bar = { urls: ["a", "b"] }
  const result = deepExtend({}, bar, foo))
```

Another question is if we even want to read `urls.primaryName` here. If we want to change it so that we only allow these types of options to be defined in the URL, we might move this to query because these lines are where we get this setting (in `master`):
https://github.com/swagger-api/swagger-ui/blob/cd7fb2778dc47ad81723ecc7013cf5aefe3405e6/src/standalone/plugins/top-bar/components/TopBar.jsx#L57-L58

https://github.com/swagger-api/swagger-ui/blob/cd7fb2778dc47ad81723ecc7013cf5aefe3405e6/src/standalone/plugins/top-bar/components/TopBar.jsx#L90

Basically: we always get the setting from `parseSearch`, so if we only want to allow this option in query, setting it here seems unnecessary.

File: src/core/config/merge.js [hunk 2]
```diff
@@ -34,6 +35,13 @@ const merge = (target, ...sources) => {
...
+    if (key.includes(".")) {
+      delete merged[key]
+      set(merged, key, value)
```
Comment (glowcloud) [inline src/core/config/merge.js:41]: In the end, I used the mutable `set` here, as we're setting `domNode` in `merged` before anyway.

Comment (char0n) [reply src/core/config/merge.js:41]: > As an FYI, the immutable version changes the order of arguments to set(key, value, merged)

Yes I aware, all functions from `/fp/` abide to functional programming immutability concepts, are auto-curried and data is always last, e.g. `set(key)(value)(merged)`.

File: src/core/config/merge.js [hunk 3]
```diff
@@ -15,25 +15,38 @@ import deepExtend from "deep-extend"
...
       domNode = source.domNode
-      const sourceWithoutDomNode = { ...source }
-      delete sourceWithoutDomNode.domNode
-      sourcesWithoutDomNode.push(sourceWithoutDomNode)
-    } else {
-      sourcesWithoutDomNode.push(source)
+      delete sourceWithoutExceptions.domNode
+    }
+
+    if (source["urls.primaryName"]) {
```
Comment (glowcloud) [inline src/core/config/merge.js:29]: Let me know if I should change these to 
```js
Object.hasOwn(source, "urls.primaryName") 
```
and
```js
Object.hasOwn(source, "urls") && Object.hasOwn(source.urls, "primaryName") 
```
for consistency with `domNode`.

Comment (char0n) [reply src/core/config/merge.js:29]: Yes please, Object.hasOwn make sure we're not dealing with inherited prototype properties.

File: src/core/config/sources/query.js [hunk 4]
```diff
@@ -9,7 +10,16 @@ import { parseSearch } from "core/utils"
  */
 
 const optionsFromQuery = () => (options) => {
```
Comment (char0n) [inline src/core/config/sources/query.js:12]: How about the following?

Pros:

- the exception of `urls.primaryName` is handled explicitly
- for any other case `set` is used
- we're using reduction on empty `{}` instead of mutating and deleting from `queryOptions`

```js
const optionsFromQuery = () => (options) => {
  const urlSearchParams = options.queryConfigEnabled ? parseSearch() : {}

  return Object.entries(urlSearchParams).reduce((acc, [key, value]) => {
    if (key === 'urls.primaryName') {
       acc[key] = value;
    } else {
       acc = set(queryOptions, key, value)
    }
    return acc;
  }, {});
}
```

File: src/core/config/merge.js [hunk 5]
```diff
@@ -15,25 +15,38 @@ import deepExtend from "deep-extend"
 
 const merge = (target, ...sources) => {
   let domNode = Symbol.for("domNode")
-  const sourcesWithoutDomNode = []
+  let primaryName = null
```
Comment (char0n) [inline src/core/config/merge.js:18]: You need to set this to something unique to be able to determine the override at the end of merge function

```suggestion
  let primaryName = Symbol.for('urls.primaryName')
```

File: src/core/config/merge.js [hunk 6]
```diff
@@ -15,25 +15,38 @@ import deepExtend from "deep-extend"
...
   }
 
-  const merged = deepExtend(target, ...sourcesWithoutDomNode)
+  const merged = deepExtend(target, ...sourcesWithoutExceptions)
 
   if (domNode !== Symbol.for("domNode")) {
     merged.domNode = domNode
   }
 
+  if (primaryName) {
```
Comment (char0n) [inline src/core/config/merge.js:46]:
```suggestion
  if (primaryName !== Symbol.for('urls.primaryName')) {
```

File: src/core/config/merge.js [hunk 7]
```diff
@@ -15,25 +15,38 @@ import deepExtend from "deep-extend"
...
+    merged.urls.primaryName = primaryName
```
Comment (char0n) [inline src/core/config/merge.js:47]: We have no way of knowing if `merged.urls` exists at this point. It's because `urls.primaryName` could have just come from `queryOptions`, but nobody bothered to set the `urls` option.

```suggestion
if (primaryName !== Symbol.for('urls.primaryName') && Array.isArray(merged.urls)) {
    merged.urls.primaryName = primaryName
}
```

File: src/core/config/merge.js [hunk 8]
```diff
@@ -15,25 +15,38 @@ import deepExtend from "deep-extend"
...
 const merge = (target, ...sources) => {
   let domNode = Symbol.for("domNode")
-  const sourcesWithoutDomNode = []
+  let primaryName = Symbol.for("primaryName")
+  const sourcesWithoutExceptions = []
 
   for (const source of sources) {
+    const sourceWithoutExceptions = { ...source }
+
     if (Object.hasOwn(source, "domNode")) {
```
Comment (char0n) [inline src/core/config/merge.js:24]:
```suggestion
    if (Object.hasOwn(sourceWithoutExceptions, "domNode")) {
```

File: src/core/config/merge.js [hunk 9]
```diff
@@ -15,25 +15,38 @@ import deepExtend from "deep-extend"
...
       domNode = source.domNode
-      const sourceWithoutDomNode = { ...source }
-      delete sourceWithoutDomNode.domNode
-      sourcesWithoutDomNode.push(sourceWithoutDomNode)
-    } else {
-      sourcesWithoutDomNode.push(source)
+      delete sourceWithoutExceptions.domNode
+    }
+
+    if (Object.hasOwn(source, "urls.primaryName")) {
```
Comment (char0n) [inline src/core/config/merge.js:29]:
```suggestion
    if (Object.hasOwn(sourceWithoutExceptions, "urls.primaryName")) {
```

File: src/core/config/merge.js [hunk 10]
```diff
@@ -15,25 +15,38 @@ import deepExtend from "deep-extend"
...
+      primaryName = source["urls.primaryName"]
+      delete sourceWithoutExceptions["urls.primaryName"]
+    } else if (source.urls && Object.hasOwn(source.urls, "primaryName")) {
```
[the remark this answers is not in the document]
Comment (char0n) [reply src/core/config/merge.js:32]:
```suggestion
    } else if (Array.isArray(sourceWithoutExceptions.urls) && Object.hasOwn(sourceWithoutExceptions.urls, "primaryName")) {
```

Comment (char0n) [inline src/core/config/merge.js:32]: Yep, your right. I've suggested changes that should make merge survive if `source` is null or undefined and compesation for `source.urls`, which we already use in https://github.com/swagger-api/swagger-ui/pull/9824/files#diff-c0a2622fbb31f1acd04a7359b8482b30f0f0eb432c92baf4eb4c2eff187f677bR46

File: src/core/config/sources/query.js [hunk 11]
```diff
@@ -1,6 +1,7 @@
 /**
  * @prettier
  */
+import set from "lodash/fp/set"
```
[the remark this answers is not in the document]
Comment (char0n) [reply src/core/config/sources/query.js:4]: Don't need immutable versions as we're reducing from the fresh object in https://github.com/swagger-api/swagger-ui/pull/9824/files#diff-0ccb6222bcef04d4fd39b85c5f1e5f2dcb4e6ce506ff9c910e8d40a4b52c8959R21, which expects to be mutated.

**Review by char0n — APPROVED** [review]
Looks good!
